### ISSUE TITLE
Add dynamic updating of Catalog when publishing APIs

### ISSUE THREAD

**Issue by JanneEN** [awslabs/aws-api-gateway-developer-portal#230]
Currently, when you publish an API in API Gateway, you need to manually export the OpenAPI file and upload it to the S3 bucket in order for the Developer Portal to update the Catalog to show the API. This manual step is cumbersome. In almost all other competing API Developer portal implementations, when an API is added or a new version of the API is updated, its description is automatically updated on the Developer Portal.

Please add automatic updates of API OpenAPI specifications to the API Gateway Developer Portal.

**Comment by a-tan** [comment]
Agreed. This will take a bit of time because we need to add a feature in API Gateway to send an event on Deployment, this will let us react to the event and update the Dev Portal.

Something to hopefully make this easier in the meantime is that we're adding an admin panel that you can use to update the APIs published with a click.

**Comment by irelanri** [comment]
I am intrigued by this thread since I cannot get ANY APIs to be listed under the Admin Panel, despite having deployed the Developer Portal both using both SAR and SAM methods, and following [the instructions](https://docs.aws.amazon.com/apigateway/latest/developerguide/apigateway-developer-portal.html#apigateway-developer-portal-publish) carefully. Is this ticket saying there is a hidden step i.e. uploading an OpenAPI file somewhere? If not, can anyone suggest any common gotchas that I may have hit?

**Comment by Trial-In-Error** [comment]
Hmm. The only step I can think of that is not documented there is to (if upgrading from a v2.x.x dev portal) delete all the existing catalog entries in the S3 bucket.

There are several things you can check:
- Did you update from a v2.x.x stack, or is this a fresh stack? Did you deploy via SAR or SAM?
- If you look in Cloudwatch Logs for the Log Stream & Group for `{stackName}-CatalogUpdaterLambdaFunction-{hash}`, do you see any errors? If so, can you paste them here? An example 'good' catalogUpdater run is available [here](https://pastebin.com/AH10M5Qj). Note that I've stripped all the stringified data out of it to make it more concise.
- If you open your browser's dev tools while signed in and on the admin panel, do you see any javascript errors? Can you provide a screenshot of the view that renders?
- If you look in S3 for the bucket whose name you provided at stack creation to put artifacts in (the recommendation in the README is `ArtifactsS3BucketName="custom-prefix-dev-portal-artifacts"`), does the bucket have a catalog.json file? What's the contents of that file?

**Comment by irelanri** [comment]
Hi, and thanks for your suggestions. This is a fresh stack and I can see that the Lambda to generate the catalog is good. However what I also see is lots of errors in the admin page of the following form:

Access to XMLHttpRequest at 'https://9nalsg6230.execute-api.eu-west-1.amazonaws.com/prod/apikey?end=undefined&sdkType=undefined&start=undefined' from origin 'https://dx1sts12wwexn.cloudfront.net' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.

catalog.json:

{"apiGateway":[{"id":"07gddq","name":"DevPortaTest","throttle":{"burstLimit":50,"rateLimit":10},"quota":{"limit":1000,"offset":0,"period":"MONTH"},"apis":[]},{"id":"0c6sl6","name":"ti_internal_ci121919_usage_plan","throttle":{"burstLimit":500,"rateLimit":50},"quota":{"limit":100000,"offset":0,"period":"DAY"},"apis":[]},{"id":"0t309f","name":"tv_master_usage_plan","throttle":{"burstLimit":500,"rateLimit":50},"quota":{"limit":100000,"offset":0,"period":"DAY"},"apis":[]},{"id":"1bd3mo","name":"tv_najla2_intercom_tv_usage_plan","throttle":{"burstLimit":500,"rateLimit":50},"quota":{"limit":10000,"offset":0,"period":"DAY"},"apis":[]},{"id":"1rd40a","name":"petstore","throttle":{"burstLimit":100,"rateLimit":10},"quota":{"limit":10000,"offset":0,"period":"MONTH"},"apis":[]},{"id":"2wgck4","name":"ti_valentfeb_usage_plan","throttle":{"burstLimit":500,"rateLimit":50},"quota":{"limit":100000,"offset":0,"period":"DAY"},"apis":[]},{"id":"3j2960","name":"ts_devios_usage_plan","throttle":{"burstLimit":50,"rateLimit":5},"quota":{"limit":1000,"offset":0,"period":"DAY"},"apis":[]},{"id":"4400pa","name":"ti_master_usage_plan","throttle":{"burstLimit":500,"rateLimit":50},"quota":{"limit":100000,"offset":0,"period":"DAY"},"apis":[]},{"id":"4ag3lc","name":"ti_devios_usage_plan","throttle":{"burstLimit":500,"rateLimit":50},"quota":{"limit":100000,"offset":0,"period":"DAY"},"apis":[]},{"id":"4g57er","name":"tv_internal_rocky2_usage_plan","throttle":{"burstLimit":500,"rateLimit":50},"quota":{"limit":10000,"offset":0,"period":"DAY"},"apis":[]},{"id":"4ji3n5","name":"ti_valentapr_usage_plan","throttle":{"burstLimit":500,"rateLimit":50},"quota":{"limit":100000,"offset":0,"period":"DAY"},"apis":[]},{"id":"4ovrw7","name":"tv_ci689644_intercom_tv_usage_plan","throttle":{"burstLimit":500,"rateLimit":50},"quota":{"limit":10000,"offset":0,"period":"DAY"},"apis":[]},{"id":"5i4c1q","name":"tv_juliob_intercom_tv_usage_plan","throttle":{"burstLimit":500,"rateLimit":50},"quota":{"limit":10000,"offset":0,"period":"DAY"},"apis":[]},{"id":"6fyppo","name":"tv_dev_usage_plan","throttle":{"burstLimit":50,"rateLimit":5},"quota":{"limit":1000,"offset":0,"period":"DAY"},"apis":[]},{"id":"6gmd1o","name":"tv_roll_intercom_tv_usage_plan","throttle":{"burstLimit":500,"rateLimit":50},"quota":{"limit":10000,"offset":0,"period":"DAY"},"apis":[]},{"id":"6h03lv","name":"tv_internal_master_usage_plan","throttle":{"burstLimit":500,"rateLimit":50},"quota":{"limit":10000,"offset":0,"period":"DAY"},"apis":[]},{"id":"6khbfn","name":"tv_juliob_usage_plan","throttle":{"burstLimit":500,"rateLimit":50},"quota":{"limit":100000,"offset":0,"period":"DAY"},"apis":[]},{"id":"73iplo","name":"tv_valentapr_usage_plan","throttle":{"burstLimit":500,"rateLimit":50},"quota":{"limit":100000,"offset":0,"period":"DAY"},"apis":[]},{"id":"7vt8zf","name":"tv_internal_valentapr_usage_plan","throttle":{"burstLimit":500,"rateLimit":50},"quota":{"limit":10000,"offset":0,"period":"DAY"},"apis":[]},{"id":"8fim1i","name":"ti_internal_ci539915_usage_plan","throttle":{"burstLimit":500,"rateLimit":50},"quota":{"limit":100000,"offset":0,"period":"DAY"},"apis":[]},{"id":"8pdxxe","name":"ti_rocky2_usage_plan","throttle":{"burstLimit":500,"rateLimit":50},"quota":{"limit":100000,"offset":0,"period":"DAY"},"apis":[]},{"id":"a3ux7i","name":"ti_deviosmars_usage_plan","throttle":{"burstLimit":500,"rateLimit":50},"quota":{"limit":100000,"offset":0,"period":"DAY"},"apis":[]},{"id":"a4xfq8","name":"tv_valentfeb_usage_plan","throttle":{"burstLimit":500,"rateLimit":50},"quota":{"limit":100000,"offset":0,"period":"DAY"},"apis":[]},{"id":"a6ab9h","name":"ts_valentfeb_usage_plan","throttle":{"burstLimit":50,"rateLimit":5},"quota":{"limit":1000,"offset":0,"period":"DAY"},"apis":[]},{"id":"ac5lrv","name":"tv_internal_devios_usage_plan","throttle":{"burstLimit":500,"rateLimit":50},"quota":{"limit":10000,"offset":0,"period":"DAY"},"apis":[]}],"generic":[]}

<img width="1437" alt="Screenshot 2019-04-04 at 14 08 27" src="https://user-images.githubusercontent.com/9928723/55558531-34360b00-56e4-11e9-8694-ce5ff7d7f503.png">

**Comment by Trial-In-Error** [comment]
Ok! The catalog is generated correctly, but it looks like there's some auth issue. The CORS messages in the dev tools are a red herring; the real issue is that you're getting a 403 from every request to the backend. Since the UI for the admin panel is visible, we know that the web application thinks you're an admin (with valid credentials), but since you're receiving 403s from the backend, the backend doesn't (possible because they've expired?).

Can you sign out and back in again and try again? If you see the same behavior after logging in and back out again, there are a couple things you could try, going from most likely to be helpful to least likely:
- create a new user, add them to the admin group, and login and see if that account works
- check the logs for the backend lambda function (named `{stackName}-DevPortalLambdaFunction-{hash}` for any 403 responses or exceptions/errors
